cranelift/codegen/src/isa/x64/inst.isle

```diff
@@ -1825,11 +1825,43 @@
 (decl extend_to_gpr (Value Type ExtendKind) Gpr)
 
 ;; If the value is already of the requested type, no extending is necessary.
+(rule 2 (extend_to_gpr (and val (value_type ty)) ty _kind)
+      val)
+
+;; I32 -> I64 with op that produces a zero-extended value in a register.
+;;
+;; As a particular x64 extra-pattern matching opportunity, all the ALU
+;; opcodes on 32-bits will zero-extend the upper 32-bits, so we can
+;; even not generate a zero-extended move in this case.
 ;;
-;; Priority 1 because the equality constraint doesn't prove that this rule
-;; doesn't overlap with the one below.
-(rule 1 (extend_to_gpr (and val (value_type ty)) ty _kind)
-      (put_in_gpr val))
+;; (Note that we unfortunately can't factor out the
+;; insts-that-zero-upper-32 pattern into a separate extractor until we
+;; can write internal extractors with multiple rules; and we'd rather
+;; keep these here than write an external extractor containing bits of
+;; the instruction patterns.)
+(rule 1 (extend_to_gpr src @ (has_type $I32 (iadd _ _)) $I64 (ExtendKind.Zero))
+        src)
+(rule 1 (extend_to_gpr src @ (has_type $I32 (isub _ _)) $I64 (ExtendKind.Zero))
+        src)
+(rule 1 (extend_to_gpr src @ (has_type $I32 (imul _ _)) $I64 (ExtendKind.Zero))
+        src)
+(rule 1 (extend_to_gpr src @ (has_type $I32 (band _ _)) $I64 (ExtendKind.Zero))
+        src)
+(rule 1 (extend_to_gpr src @ (has_type $I32 (bor _ _)) $I64 (ExtendKind.Zero))
+        src)
+(rule 1 (extend_to_gpr src @ (has_type $I32 (bxor _ _)) $I64 (ExtendKind.Zero))
+        src)
+(rule 1 (extend_to_gpr src @ (has_type $I32 (ishl _ _)) $I64 (ExtendKind.Zero))
+        src)
+(rule 1 (extend_to_gpr src @ (has_type $I32 (ushr _ _)) $I64 (ExtendKind.Zero))
+        src)
+(rule 1 (extend_to_gpr src @ (has_type $I32 (uload32 _ _ _)) $I64 (ExtendKind.Zero))
+        src)
+
+;; The `pextr{b,w}` instruction automatically zero the upper bits of the
+;; destination register so the `uextend` in these cases can be elided.
+(rule 1 (extend_to_gpr src @ (has_type (fits_in_16 _) (extractlane _ _)) $I32 (ExtendKind.Zero))
```

Comment:
This seems like a nice optimization for `extractlane` but I'm concerned that it may be fragile. Correct me if I'm missing something, but I think it's relying on the lowering rules for `extractlane` to produce specific instructions, without having any clear connection between the two sets of rules. Not only are they far removed from each other, but this rule checks that the result type `fits_in_16`, while that rule checks that the argument type is `multi_lane 8` or `multi_lane 16`.

At minimum I think each place needs a comment referring to the other. Ideally it'd be great to connect them in some way that's harder to get wrong, though I haven't come up with a good way to do that.

Reply:
True!

One way this could be tackled is to just go ahead and emit the instruction right here in the helper. For the case of `extractlane` that's pretty easy since it's a single instruction and the `extractlane` rules themselves aren't fancy. That's still somewhat brittle though as it doesn't handle evolution of `extractlane` over time (if at all).

Perhapes more worrisome, though, is the preexisting cases for dropping a `uextend`. The pattern matches for 32-bit I think have the same trait where they don't necessarily guarantee a particular instruction is being used, we're just relying on things being implemented in a particular way. That part I don't know how best to solve because I definitely wouldn't want to inline all the `iadd` rules here, for example.

Do you think it's worthwhile to comment all of these though? While perhaps exhaustive it may be getting a bit out of hand since a good handful of the binops span multiple terminals of rules and if the purpose of the comment is to say "please always use the `sub` intruction here or something else that zeros the upper 32-bits for `uextend` elision" then it's easy to miss that if it's only located at the top of the rules.

I suppose an even crazier strategy would be to postprocess the instructions after lowering, but that naturally has no existing infrastructure and is almost certainly overkill for this one use case.

Reply:
Oh, I guess I assumed that we could pretty well count on any implementation of e.g. 32-bit add having this behavior on x86. I thought of `pextrb` as a little more magical because it zero-extends to a narrower width.

I'm not saying this is a good plan, but I guess another approach would be to scatter the individual `extend_to_gpr` and `value32_zeros_upper32` rules out to place them next to the `lower` rules that they rely on. Actually, I don't entirely hate that.

We don't necessarily have to solve this in this PR, I'm just a little nervous about it long-term. I am curious if @cfallin or others have any other suggestions for this.

Reply:
Yeah, unfortunately this is the status-quo for `iadd` and friends too. Placing rules that are intertwined in this way (producer that zeroes upper bits, consumer that relies on it) next to each other is not a bad idea; or at least, we could add comments nothing the implicit extra invariant everywhere it is either enforced or relied upon.

At various points we've talked about "demanded-bits" and "known-bits" optimizations that could solve this sort of issue at the lowering-framework level: for example, in the upward scan during isel, we could record "requires zero-extended value" when we see a uextend and otherwise pass through the source register, then generate the extend after the *producer* if any consumer needed it. That needs a bit more thought to refine the right abstractions though.

We could also optimize VCode; at various points we've talked about post-lowering GVN, cprop, etc., and eliding redundant uextends could be a part of that. But again, much more plumbing needed.

Finally, this is exactly the sort of invariant that the verification project encodes in annotations, and checks on both ends. So I'm optimistic that *eventually* we'll be in a situation where it's actually enforced, and we can sleep easy at night. In the meantime IMHO it's probably enough to rely on documentation via comments and/or placing intertwined rules together.

Reply:
I should note that the [`pextrb` documentation](https://www.felixcloutier.com/x86/pextrb:pextrd:pextrq) I'm pretty sure indicates that it zero-extends to the whole 64-bit register, but I'm not certain enough to add the rule just yet and otherwise trying to write manual tests for this is a bit of a pain. I'm also mostly interested in this from a wasm perspective where [`i8x16.extract_lane_u` for example](https://github.com/bytecodealliance/wasmtime/blob/5ae8575296d5b524cde42ad10badf8c89945105a/cranelift/wasm/src/code_translator.rs#L1638-L1641) unconditionally does a `uextend` to `i32` to obey wasm semantics but the `uextend` can be elided in all cases for the x86 instruction (and probably other platforms too which I should look into).

One point of hesitation I would have to scatter the `(rule ...)`s to various places would be that priorities are needed for the `(rule ...)` annotations and it's easy enough to reason about priorities when they're written in one place but having to group them from lots of different places could make things difficult. 

Given though that the main motivation is the wasm lowering here, I could also just add this one specific rule for `uextend` around the place that `extractlane` is codegen'd since they'd both use the single `pextr{b,w}` instruction.

I'm still not sure how best to resolve this, but if you're mainly worried about the `extractlane` cases being the odd ones out, how about I add comments to the `(rule)` with extractlane here and comments on the lowering of `extractlane` as well pointing to each other?

Reply:
I think adding comments just for extractlane is plenty for this PR. Thanks!

I agree with your concerns about scattering rules around that rely on priorities. I think we can deal with that, and also I think all our options are bad in different ways, which is why I said I don't hate that option. But in any case it doesn't have to happen right now.

Reply:
Ok I have added some more words to help connect the dots here 👍